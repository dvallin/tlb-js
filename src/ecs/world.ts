import { Storage } from './storage'
import { System } from './system'
import { Resource } from './resource'
import { Entity, EntityModifier } from './entity'

export class World<C, S, R> {
  public readonly components: Map<C, Storage<object>> = new Map()
  public readonly resources: Map<R, Resource<C, S, R>> = new Map()
  public readonly systems: Map<S, System<C, S, R>> = new Map()
  public readonly emptySystems: Set<S> = new Set()
  public readonly activeSystems: Set<S> = new Set()

  private openEntities: Set<Entity> = new Set()
  private lastEntity: Entity = -1

  public get entityCount(): number {
    return this.lastEntity + 1 - this.openEntities.size
  }

  public registerComponentStorage<T extends object>(component: C, storage: Storage<T>): void {
    this.components.set(component, storage)
  }

  public getStorage<T extends object>(component: C): Storage<T> {
    return this.components.get(component)! as Storage<T>
  }

  public getResource<T extends Resource<C, S, R>>(resource: R): T {
    return this.resources.get(resource)! as T
  }

  public getComponent<T extends object>(entity: Entity, component: C): T | undefined {
    const storage = this.getStorage<T>(component)
    if (storage !== undefined) {
      return storage.get(entity)
    }
    return undefined
  }

  public hasEntity(entity: Entity): boolean {
    return entity <= this.lastEntity && !this.openEntities.has(entity)
  }

  public hasComponent(entity: Entity, component: C): boolean {
    const storage = this.getStorage(component)
    if (storage !== undefined) {
      return storage.has(entity)
    }
    return false
  }

  public createEntity(): EntityModifier<C, S, R> {
    let entity
    if (this.openEntities.size > 0) {
      entity = this.openEntities.values().next().value
      this.openEntities.delete(entity)
    } else {
      entity = ++this.lastEntity
    }
    return new EntityModifier(this, entity)
  }

  public editEntity(entity: Entity): EntityModifier<C, S, R> {
    return new EntityModifier(this, entity)
  }

  public deleteEntity(entity: Entity): void {
    this.editEntity(entity).delete()
    this.openEntities.add(entity)
  }

  public registerSystem(name: S, system: System<C, S, R>): void {
    this.systems.set(name, system)
  }

  public enableSystem(name: S): void {
    this.activeSystems.add(name)
  }

  public disableSystem(name: S): void {
    this.activeSystems.delete(name)
  }

  public activeSystemsList(): S[] {
    const l: S[] = []
    this.activeSystems.forEach(s => l.push(s))
    return l
  }

  public registerResource(resource: Resource<C, S, R>): void {
    this.resources.set(resource.kind, resource)
  }

  public updateResources(): void {
    for (const resource of this.resources) {
      resource[1].update(this)
    }
  }

  public updateSystems(): void {
    this.emptySystems.clear()
    for (const name of this.activeSystems) {
      const system = this.systems.get(name)!
      const firstStorage = this.getStorage(system.components[0])
      let calls = 0
      firstStorage.foreach((entity: Entity) => {
        let callSystem = true
        for (let i = 1; i < system.components.length; i++) {
          if (!this.hasComponent(entity, system.components[i])) {
            callSystem = false
            break
          }
        }
        if (callSystem) {
          system.update(this, entity)
          calls++
        }
      })
      if (calls === 0) {
        this.emptySystems.add(name)
      }
    }
  }
}
