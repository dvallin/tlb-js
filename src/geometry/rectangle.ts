import { AbstractShape } from './shape'
import { Vector } from '../spatial'
import { Direction } from '../spatial/direction'

export class Rectangle extends AbstractShape {
  public static fromBounds(left: number, right: number, top: number, bottom: number): Rectangle {
    return new Rectangle(left, top, right - left + 1, bottom - top + 1)
  }

  public static footprint(position: Vector, direction: Direction, size: Vector): Rectangle {
    const up: Vector = Vector.fromDirection(direction).abs()
    const right: Vector = Vector.fromDirection(direction)
      .perpendicular()
      .abs()
    const cx = Math.floor(size.x / 2)
    const cy = Math.floor(size.y / 2)
    const topLeft = position.add(right.mult(-cx)).add(up.mult(-cy))
    const bottomRight = topLeft.add(right.mult(size.x - 1)).add(up.mult(size.y - 1))
    return Rectangle.fromBounds(topLeft.x, bottomRight.x, topLeft.y, bottomRight.y)
  }

  public static centerAt(x: number, y: number, size: number): Rectangle {
    return new Rectangle(x - size, y - size, 2 * size + 1, 2 * size + 1)
  }

  public constructor(public readonly x: number, public readonly y: number, public readonly width: number, public readonly height: number) {
    super()
  }

  public get left(): number {
    return this.x
  }

  public get right(): number {
    return this.x + this.width - 1
  }

  public get top(): number {
    return this.y
  }

  public get bottom(): number {
    return this.y + this.height - 1
  }

  public get topLeft(): Vector {
    return new Vector([this.left, this.top])
  }

  public get topRight(): Vector {
    return new Vector([this.right, this.top])
  }

  public get bottomRight(): Vector {
    return new Vector([this.right, this.bottom])
  }

  public get bottomLeft(): Vector {
    return new Vector([this.left, this.bottom])
  }

  public get center(): Vector {
    const cx = (this.left + this.right) / 2
    const cy = (this.top + this.bottom) / 2
    return new Vector([Math.floor(cx), Math.floor(cy)])
  }

  public centerOf(direction: Direction): Vector {
    switch (direction) {
      case 'up':
        return this.centerTop
      case 'right':
        return this.centerRight
      case 'down':
        return this.centerBottom
      case 'left':
        return this.centerLeft
    }
  }

  public get centerLeft(): Vector {
    const cy = (this.top + this.bottom) / 2
    return new Vector([this.left, Math.floor(cy)])
  }

  public get centerRight(): Vector {
    const cy = (this.top + this.bottom) / 2
    return new Vector([this.right, Math.floor(cy)])
  }

  public get centerTop(): Vector {
    const cx = (this.left + this.right) / 2
    return new Vector([Math.floor(cx), this.top])
  }

  public get centerBottom(): Vector {
    const cx = (this.left + this.right) / 2
    return new Vector([Math.floor(cx), this.bottom])
  }

  public plus(other: Rectangle): Rectangle {
    return Rectangle.fromBounds(
      Math.min(this.left, other.left),
      Math.max(this.right, other.right),
      Math.min(this.top, other.top),
      Math.max(this.bottom, other.bottom)
    )
  }

  public intersect(other: Rectangle): Rectangle {
    return Rectangle.fromBounds(
      Math.max(this.left, other.left),
      Math.min(this.right, other.right),
      Math.max(this.top, other.top),
      Math.min(this.bottom, other.bottom)
    )
  }

  public cover(other: Vector): Rectangle {
    return Rectangle.fromBounds(
      Math.min(this.left, other.x),
      Math.max(this.right, other.x),
      Math.min(this.top, other.y),
      Math.max(this.bottom, other.y)
    )
  }

  public bounds(): Rectangle {
    return this
  }

  public containsVector(p: Vector): boolean {
    return p.x >= this.x && p.x < this.x + this.width && p.y >= this.y && p.y < this.y + this.height
  }

  public translate(t: Vector): Rectangle {
    return new Rectangle(this.x + t.x, this.y + t.y, this.width, this.height)
  }

  public grow(cells: number = 1): Rectangle {
    return new Rectangle(this.x - cells, this.y - cells, this.width + 2 * cells, this.height + 2 * cells)
  }

  public shrink(cells: number = 1): Rectangle {
    const w = this.width - 2 * cells
    const h = this.height - 2 * cells
    return new Rectangle(w > 0 ? this.x + cells : this.x, h > 0 ? this.y + cells : this.y, w > 0 ? w : 0, h > 0 ? h : 0)
  }

  public all(f: (position: Vector) => boolean): boolean {
    for (let i = 0; i < this.height; i++) {
      for (let j = 0; j < this.width; j++) {
        const position = new Vector([j + this.x, i + this.y])
        if (!f(position)) {
          return false
        }
      }
    }
    return true
  }
}
