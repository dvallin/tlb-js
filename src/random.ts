import * as seedrandom from "seedrandom"

let rng: seedrandom.prng

export function init(seed?: string): void {
    rng = seedrandom(seed)
}

export function uniformInteger(min: number = 0, max: number = 1): number {
    return (Math.abs(rng.int32()) % (max - min)) + min
}

export function bernoulli(p: number = 0.5): boolean {
    return rng.double() <= p
}

export function binomialNormed(p: number[]): number {
    let test = rng.double()
    return p.findIndex((t) => {
        if (test <= t) {
            return true
        }
        test -= t
        return false
    })
}

export function dualDecision(
    p00: number, [p01, p10]: number[], p11: number
): [boolean, boolean] {
    const decision = binomialNormed([p00, p01, p10, p11])
    return [
        Math.floor(decision / 2) % 2 === 1,
        decision % 2 === 1
    ]
}

export function ternaryDecision(
    p000: number, [p001, p010, p100]: number[], [p011, p101, p110]: number[], p111: number
): [boolean, boolean, boolean] {
    const decision = binomialNormed([p000, p001, p010, p011, p100, p101, p110, p111])
    return [
        Math.floor(decision / 4) % 2 === 1,
        Math.floor(decision / 2) % 2 === 1,
        decision % 2 === 1
    ]
}
