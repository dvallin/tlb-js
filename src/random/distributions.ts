import { create, RandomSeed } from "random-seed"

export interface Distribution {
    sample(): number
}

export class Uniform implements Distribution {
    public readonly rng: RandomSeed

    public constructor(seed: string) {
        this.rng = create(seed)
    }

    public sample(): number {
        return this.rng.random()
    }
}

export class Exponential implements Distribution {

    public static fromSeed(seed: string, lambda: number = 1) {
        return new Exponential(new Uniform(seed), lambda)
    }

    private scale: number

    public constructor(
        public readonly distribution: Distribution,
        public readonly lambda: number = 1
    ) {
        this.scale = 1 - Math.exp(-this.lambda)
    }

    public sample(): number {
        const u = this.distribution.sample()
        return -Math.log(1 - this.scale * u) / this.lambda
    }
}
