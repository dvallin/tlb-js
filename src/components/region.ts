import { Shape } from '../geometry/shape'
import { Vector } from '../spatial'
import { Entity } from '../ecs/entity'

export type RegionsType = 'red'

export interface RegionComponent {
  type: RegionsType
  level: number
  shape: Shape
  entry: Vector | undefined
  authorized: Set<Entity>
}

export type StructureType = 'room' | 'hub' | 'corridor'

export interface StructureComponent {
  region: Entity
  kind: StructureType
  shape: Shape
  connections: Entity[]
}
