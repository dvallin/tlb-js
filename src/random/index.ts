import { Distribution } from './distributions'
import { Rectangle } from '../geometry/rectangle'
import { Vector } from '../spatial'

export class Random {
  public constructor(public readonly distribution: Distribution) {}

  public floatBetween(minInclusive: number, maxExclusive: number): number {
    return this.distribution.sample() * (maxExclusive - minInclusive) + minInclusive
  }

  public integerBetween(minInclusive: number, maxInclusive: number): number {
    const c = this.distribution.sample()
    return Math.floor(c * (maxInclusive - minInclusive + 1)) + minInclusive
  }

  public decision(probability: number): boolean {
    return this.distribution.sample() < probability
  }

  public weightedDecision(weights: number[]): number {
    const sum = weights.reduce((v, c) => v + c, 0)
    const pick = this.integerBetween(0, sum - 1)
    let agg = 0
    for (let i = 0; i < weights.length; i++) {
      agg += weights[i]
      if (pick < agg) {
        return i
      }
    }
    throw new Error('invalid input to weighted decision')
  }

  public pick<T>(array: T[]): T {
    return array[this.integerBetween(0, array.length - 1)]
  }

  public pickIndex<T>(array: T[]): number {
    return this.integerBetween(0, array.length - 1)
  }

  public shuffle<T>(array: T[]): void {
    for (let i = array.length - 1; i >= 0; i--) {
      const j = this.integerBetween(0, i)
      const t = array[i]
      array[i] = array[j]
      array[j] = t
    }
  }

  public insideRectangle(rectangle: Rectangle): Vector {
    return new Vector([this.integerBetween(rectangle.left, rectangle.right), this.integerBetween(rectangle.top, rectangle.bottom)])
  }
}
