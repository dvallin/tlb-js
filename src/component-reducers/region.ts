import { TlbWorld } from '../tlb'
import { StructureComponent, RegionComponent } from '../components/region'
import { Entity } from '../ecs/entity'
import { PositionComponent } from '../components/position'
import { WorldMap, WorldMapResource } from '../resources/world-map'

export function isAuthorized(world: TlbWorld, position: PositionComponent, entity: Entity): boolean {
  const map: WorldMap = world.getResource<WorldMapResource>('map')
  const structure = map.levels[position.level].getStructure(position.position)!
  const room: StructureComponent = world.getComponent<StructureComponent>(structure, 'structure')!
  const region: RegionComponent = world.getComponent<RegionComponent>(room.region, 'region')!
  return region.authorized.has(entity)
}

export function authorize(world: TlbWorld, position: PositionComponent, entity: Entity): void {
  const map: WorldMap = world.getResource<WorldMapResource>('map')
  const structure = map.levels[position.level].getStructure(position.position)!
  const room: StructureComponent = world.getComponent<StructureComponent>(structure, 'structure')!
  const region: RegionComponent = world.getComponent<RegionComponent>(room.region, 'region')!
  region.authorized.add(entity)
}
