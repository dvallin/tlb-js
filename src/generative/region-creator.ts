import { TlbWorld } from '../tlb'
import { RegionComponent, RegionsType } from '../components/region'
import { Rectangle } from '../geometry/rectangle'
import { Shape } from '../geometry/shape'
import { Entity } from '../ecs/entity'
import { Vector } from '../spatial'

function regionComponent(type: RegionsType, shape: Shape, exits: Vector[], level: number): RegionComponent {
  return { type, shape, level, exits, authorized: new Set() }
}

export function create(world: TlbWorld): Entity {
  const elevatorRegion = new Rectangle(50, 25, 20, 20)
  world
    .createEntity()
    .withComponent<RegionComponent>('region', regionComponent('elevator', elevatorRegion, [new Vector([50, 28])], 0))
    .withComponent('active', {})
  world
    .createEntity()
    .withComponent<RegionComponent>('region', regionComponent('elevator', elevatorRegion, [new Vector([50, 28])], 1))
    .withComponent('active', {})
  world
    .createEntity()
    .withComponent<RegionComponent>('region', regionComponent('red', new Rectangle(0, 0, 50, 50), [new Vector([49, 28])], 1))
    .withComponent('active', {}).entity
  return world
    .createEntity()
    .withComponent<RegionComponent>('region', regionComponent('red', new Rectangle(0, 0, 50, 50), [new Vector([49, 28])], 0))
    .withComponent('active', {}).entity
}
