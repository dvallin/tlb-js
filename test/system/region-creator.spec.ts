import { RegionCreator } from '../../src/systems/region-creator'
import { Uniform } from '../../src/random/distributions'
import { World } from '../../src/ecs/world'
import { TlbWorld, registerComponents, registerResources } from '../../src/tlb'
import { RegionComponent } from '../../src/components/region'
import { Rectangle } from '../../src/geometry/rectangle'
import { Vector } from '../../src/spatial'
import { mockRenderer } from '../mocks'
import { renderMap } from '../render'

describe('RegionCreator', () => {
  it('creates a region', () => {
    const world: TlbWorld = new World()
    registerComponents(world)
    registerResources(world, mockRenderer())
    const shape = new Rectangle(0, 0, 30, 30)
    const region = world
      .createEntity()
      .withComponent('active', {})
      .withComponent<RegionComponent>('region', {
        type: 'red',
        shape,
        level: 0,
        authorized: new Set(),
        entry: new Vector([0, 0]),
      }).entity

    const system = new RegionCreator(new Uniform('12'))

    system.update(world, region)

    expect(renderMap(world, shape)).toMatchSnapshot()
  })
})
