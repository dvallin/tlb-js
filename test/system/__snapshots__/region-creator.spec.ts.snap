// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegionCreator creates a region 1`] = `
"                    ∏..       
 ......... .....gg. .╔╗ ..... 
 ......... ........ .╚╝ ...g. 
 ......... ........ .╔╗◊..... 
 ......... ........ .╚╝◊..... 
 ......... ........ ...◊.... 
 .........◊.....g.. ... ..... 
 .....g...◊........ ... ..... 
 .........◊........ ...  ◊◊◊  
 ......... ..g..... ... .... 
 ......... ........ ╔╗.◊..... 
 ......... ........ ╚╝.◊..... 
 ......... ........ ...◊..... 
 ......... g....... ... ...g. 
    ◊◊◊      ◊◊◊    ...  ◊◊◊  
..............................
..............................
..............................
           ...      ◊◊◊       
 ......... .╔╗ .............. 
 ......... .╚╝◊.............. 
 ......... ...◊.............. 
 ......... ...◊.............. 
 ......... ... .............. 
    ◊◊◊    .╔╗      ◊◊◊       
 .........◊.╚╝◊.............. 
 .........◊...◊.............. 
 .........◊...◊.............. 
 ......... .╔╗ .............. 
           .╚╝                "
`;
